shap/_explanation.py

```diff
@@ -826,33 +826,39 @@ def _compute_shape(x) -> tuple[Union[int, None], ...]:
 
 
 class Cohorts:
-    def __init__(self, **kwargs):
+    def __init__(self, **kwargs) -> None:
         self.cohorts = kwargs
-        for k in self.cohorts:
-            assert isinstance(
-                self.cohorts[k], Explanation
-            ), "All the arguments to a Cohorts set must be Explanation objects!"
 
-    def __getitem__(self, item):
-        new_cohorts = Cohorts()
-        for k in self.cohorts:
-            new_cohorts.cohorts[k] = self.cohorts[k].__getitem__(item)
-        return new_cohorts
-
-    def __getattr__(self, name):
+    @property
+    def cohorts(self):
+        return self._cohorts
+
+    @cohorts.setter
+    def cohorts(self, cval):
+        if not isinstance(cval, dict):
+            emsg = "self.cohorts must be a dictionary!"
+            raise TypeError(emsg)
+        if not all(isinstance(exp, Explanation) for exp in cval.values()):
+            emsg = "All the arguments to a Cohorts set must be Explanation objects!"
+            raise ValueError(emsg)
+
+        cast(dict[str, Explanation], cval)  # type narrowing for mypy
+        self._cohorts: dict[str, Explanation] = cval
+
+    def __getitem__(self, item) -> "Cohorts":
         new_cohorts = Cohorts()
-        for k in self.cohorts:
-            new_cohorts.cohorts[k] = getattr(self.cohorts[k], name)
+        for k in self._cohorts:
+            new_cohorts.cohorts[k] = self._cohorts[k].__getitem__(item)
         return new_cohorts
 
-    def __call__(self, *args, **kwargs):
+    def __getattr__(self, name) -> "Cohorts":
         new_cohorts = Cohorts()
-        for k in self.cohorts:
-            new_cohorts.cohorts[k] = self.cohorts[k].__call__(*args, **kwargs)
+        for k in self._cohorts:
+            new_cohorts.cohorts[k] = getattr(self._cohorts[k], name)
         return new_cohorts
```

Comment:
`__call__()` is removed here, it makes no sense once we enforce that `self.cohorts[k]` must be an Explanation object.

Reply:
NOTE to self: The above claim is somewhat wrong. We still need to retain `__call__` here in order to invoke any methods that were retrieved during `__getattr__`.

E.g. `Cohorts(...).mean(axis=0)` would first run `__getattr__("mean")` and return a bound method `Explanation.mean` for all the Explanation objects inside the Cohorts. And return a Cohorts object.
Then the `(axis=0)` call would be executed on that returned Cohorts object, which is why we need `__call__`.

Reply:
Nice! I think it would help to include your explanation above as a comment in the `__call__` method, as it's quite subtle.

Reply:
done